File: modulo01-fundamentos-de-ia-e-llms-para-programadores/ecommerce-neural-recommender-java/src/main/java/com/recomendacao/ml/ModelContext.java
package com.recomendacao.ml;

import com.recomendacao.model.Product;
import com.recomendacao.model.User;

import java.util.*;

/**
 * Contexto do modelo: contém os metadados de codificacao calculados a partir
 * dos dados de treino. Equivalente ao objeto retornado por makeContext() no JS.
 *
 * Guarda:
 *  - Limites de normalizacao (minPrice, maxPrice, minAge, maxAge)
 *  - Indices para one-hot encoding de categorias e cores
 *  - Idade media normalizada por produto (para personalizacao)
 *  - Vetores pre-calculados de cada produto
 */
public class ModelContext {

    // Pesos de cada feature na composicao do vetor
    // (espelho do objeto WEIGHTS do JS)
    public static final double WEIGHT_CATEGORY = 0.4;
    public static final double WEIGHT_COLOR    = 0.3;
    public static final double WEIGHT_PRICE    = 0.2;
    public static final double WEIGHT_AGE      = 0.1;

    public final List<Product> products;
    public final List<User>    users;

    public final double minAge;
    public final double maxAge;
    public final double minPrice;
    public final double maxPrice;

    // Mapeamento: nome -> indice para one-hot encoding
    public final Map<String, Integer> categoriesIndex;
    public final Map<String, Integer> colorsIndex;

    public final int numCategories;
    public final int numColors;

    // dimensoes = price(1) + age(1) + categorias + cores
    public final int dimensions;

    // Idade media normalizada de compradores por produto
    // (ajuda a personalizar por faixa etaria)
    public final Map<String, Double> productAvgAgeNorm;

    // Vetores pre-calculados de cada produto para evitar reprocessamento
    // durante a recomendacao
    public double[][] productVectors;

    public ModelContext(List<Product> products, List<User> users) {
        this.products = products;
        this.users    = users;

        // --- Calcular limites de normalizacao ---
        this.minAge   = users.stream().mapToInt(User::getAge).min().orElse(0);
        this.maxAge   = users.stream().mapToInt(User::getAge).max().orElse(100);
        this.minPrice = products.stream().mapToDouble(Product::getPrice).min().orElse(0);
        this.maxPrice = products.stream().mapToDouble(Product::getPrice).max().orElse(1000);

        // --- Construir indices para one-hot encoding ---
        List<String> categories = products.stream()
                .map(Product::getCategory).distinct().sorted().toList();
        List<String> colors = products.stream()
                .map(Product::getColor).distinct().sorted().toList();

        this.categoriesIndex = new LinkedHashMap<>();
        for (int i = 0; i < categories.size(); i++) categoriesIndex.put(categories.get(i), i);

        this.colorsIndex = new LinkedHashMap<>();
        for (int i = 0; i < colors.size(); i++) colorsIndex.put(colors.get(i), i);

        this.numCategories = categories.size();
        this.numColors     = colors.size();
        this.dimensions    = 2 + numCategories + numColors;

        // --- Calcular idade media normalizada por produto ---
        Map<String, Double> ageSums   = new HashMap<>();
        Map<String, Integer> ageCounts = new HashMap<>();

        for (User user : users) {
            for (Product p : user.getPurchases()) {
                ageSums.merge(p.getName(), (double) user.getAge(), Double::sum);
                ageCounts.merge(p.getName(), 1, Integer::sum);
            }
        }

        double midAge = (minAge + maxAge) / 2.0;
        this.productAvgAgeNorm = new HashMap<>();
        for (Product product : products) {
            double avg = ageCounts.containsKey(product.getName())
                    ? ageSums.get(product.getName()) / ageCounts.get(product.getName())
                    : midAge;
            productAvgAgeNorm.put(product.getName(), normalize(avg, minAge, maxAge));
        }
    }

    /**
     * Normaliza um valor para o intervalo [0, 1].
     * Formula: (val - min) / (max - min)
     * Equivalente a normalize() do JS.
     */
    public static double normalize(double value, double min, double max) {
        double range = max - min;
        return range == 0 ? 0 : (value - min) / range;
    }
}

File: modulo01-fundamentos-de-ia-e-llms-para-programadores/ecommerce-neural-recommender-go/ml/model_context.go
package ml

import (
	"sort"

	"ecommerce-neural-recommender/model"
)

// Pesos de cada feature na composição do vetor — espelho do objeto WEIGHTS
// da versão Java/JS.
const (
	WeightCategory = 0.4
	WeightColor    = 0.3
	WeightPrice    = 0.2
	WeightAge      = 0.1
)

// ModelContext contém os metadados de codificação calculados a partir dos
// dados de treino: limites de normalização, índices de one-hot encoding e
// idade média normalizada por produto — equivalente a ModelContext.java.
type ModelContext struct {
	Products []model.Product
	Users    []model.User

	MinAge, MaxAge     float64
	MinPrice, MaxPrice float64

	CategoriesIndex map[string]int
	ColorsIndex     map[string]int

	NumCategories int
	NumColors     int

	// Dimensions = price(1) + age(1) + categorias + cores.
	Dimensions int

	// Idade média normalizada de compradores por produto.
	ProductAvgAgeNorm map[string]float64

	// Vetores pré-calculados de cada produto (preenchidos pelo
	// RecommendationEngine para evitar reprocessamento).
	ProductVectors [][]float64
}

// NewModelContext calcula todos os metadados de codificação a partir dos
// produtos e usuários informados.
func NewModelContext(products []model.Product, users []model.User) *ModelContext {
	ctx := &ModelContext{Products: products, Users: users}

	ctx.MinAge, ctx.MaxAge = ageBounds(users)
	ctx.MinPrice, ctx.MaxPrice = priceBounds(products)

	categories := distinctSorted(products, func(p model.Product) string { return p.Category })
	colors := distinctSorted(products, func(p model.Product) string { return p.Color })

	ctx.CategoriesIndex = indexOf(categories)
	ctx.ColorsIndex = indexOf(colors)
	ctx.NumCategories = len(categories)
	ctx.NumColors = len(colors)
	ctx.Dimensions = 2 + ctx.NumCategories + ctx.NumColors

	ctx.ProductAvgAgeNorm = productAvgAgeNorm(products, users, ctx.MinAge, ctx.MaxAge)

	return ctx
}

// Normalize normaliza um valor para o intervalo [0, 1]: (val - min) / (max - min).
func Normalize(value, min, max float64) float64 {
	rangeVal := max - min
	if rangeVal == 0 {
		return 0
	}
	return (value - min) / rangeVal
}

func ageBounds(users []model.User) (min, max float64) {
	if len(users) == 0 {
		return 0, 100
	}
	min, max = float64(users[0].Age), float64(users[0].Age)
	for _, u := range users[1:] {
		age := float64(u.Age)
		if age < min {
			min = age
		}
		if age > max {
			max = age
		}
	}
	return min, max
}

func priceBounds(products []model.Product) (min, max float64) {
	if len(products) == 0 {
		return 0, 1000
	}
	min, max = products[0].Price, products[0].Price
	for _, p := range products[1:] {
		if p.Price < min {
			min = p.Price
		}
		if p.Price > max {
			max = p.Price
		}
	}
	return min, max
}

func distinctSorted(products []model.Product, key func(model.Product) string) []string {
	seen := map[string]bool{}
	var out []string
	for _, p := range products {
		k := key(p)
		if !seen[k] {
			seen[k] = true
			out = append(out, k)
		}
	}
	sort.Strings(out)
	return out
}

func indexOf(values []string) map[string]int {
	idx := make(map[string]int, len(values))
	for i, v := range values {
		idx[v] = i
	}
	return idx
}

func productAvgAgeNorm(products []model.Product, users []model.User, minAge, maxAge float64) map[string]float64 {
	ageSums := map[string]float64{}
	ageCounts := map[string]int{}

	for _, user := range users {
		for _, p := range user.Purchases {
			ageSums[p.Name] += float64(user.Age)
			ageCounts[p.Name]++
		}
	}

	midAge := (minAge + maxAge) / 2.0
	result := make(map[string]float64, len(products))
	for _, product := range products {
		var avg float64
		if count, ok := ageCounts[product.Name]; ok {
			avg = ageSums[product.Name] / float64(count)
		} else {
			avg = midAge
		}
		result[product.Name] = Normalize(avg, minAge, maxAge)
	}
	return result
}
